From 6db72a7adb2f2e370056b44cfed00b49364c1dcf Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
Date: Wed, 19 Nov 2025 16:19:36 +0800
Subject: [PATCH 126/128] mtk: mac80211: add support for emlsr enablement on
 one link

Add support for emlsr enablement on one link

Signed-off-by: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
---
 include/net/cfg80211.h       | 2 ++
 include/uapi/linux/nl80211.h | 2 ++
 net/wireless/nl80211.c       | 7 ++++++-
 3 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index ffa3172b..8d5f0970 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -5666,6 +5666,7 @@ struct wiphy_vendor_command {
  * @extended_capabilities_len: length of the extended capabilities
  * @eml_capabilities: EML capabilities (for MLO)
  * @mld_capa_and_ops: MLD capabilities and operations (for MLO)
+ * @extended_mld_capa_and_ops: Extended MLD capabilities and Operations (for MLO)
  */
 struct wiphy_iftype_ext_capab {
 	enum nl80211_iftype iftype;
@@ -5674,6 +5675,7 @@ struct wiphy_iftype_ext_capab {
 	u8 extended_capabilities_len;
 	u16 eml_capabilities;
 	u16 mld_capa_and_ops;
+	u16 ext_mld_capa_and_ops;
 };
 
 /**
diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index bf2eca25..353ec69d 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -2860,6 +2860,7 @@ enum nl80211_commands {
  *
  * @NL80211_ATTR_EML_CAPABILITY: EML Capability information (u16)
  * @NL80211_ATTR_MLD_CAPA_AND_OPS: MLD Capabilities and Operations (u16)
+ * @NL80211_ATTR_EXT_MLD_CAPA_AND_OPS: Extended MLD Capabilities and Operations (u16)
  *
  * @NL80211_ATTR_TX_HW_TIMESTAMP: Hardware timestamp for TX operation in
  *	nanoseconds (u64). This is the device clock timestamp so it will
@@ -3607,6 +3608,7 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_DFS_STATE_UPDATE,
 	NL80211_ATTR_BROADCAST_TWT_PARAMS,
+	NL80211_ATTR_EXT_MLD_CAPA_AND_OPS,
 
 	__NL80211_ATTR_AFTER_LAST,
 	NUM_NL80211_ATTR = __NL80211_ATTR_AFTER_LAST,
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index b8dcaafa..c1bc0530 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -981,6 +981,8 @@ static const struct nla_policy nl80211_policy[NUM_NL80211_ATTR] = {
 	[NL80211_ATTR_VENDOR_MTK_STA] = { .type = NLA_FLAG },
 	[NL80211_ATTR_BROADCAST_TWT_PARAMS] =
 		NLA_POLICY_NESTED(nl80211_broadcast_twt_policy),
+	[NL80211_ATTR_MLD_CAPA_AND_OPS] = { .type = NLA_U16 },
+	[NL80211_ATTR_EXT_MLD_CAPA_AND_OPS] = { .type = NLA_U16 },
 };
 
 /* policy for the key attributes */
@@ -3249,7 +3251,10 @@ static int nl80211_send_wiphy(struct cfg80211_registered_device *rdev,
 						 capab->eml_capabilities) ||
 				     nla_put_u16(msg,
 						 NL80211_ATTR_MLD_CAPA_AND_OPS,
-						 capab->mld_capa_and_ops)))
+						 capab->mld_capa_and_ops) ||
+				     nla_put_u16(msg,
+						 NL80211_ATTR_EXT_MLD_CAPA_AND_OPS,
+						 capab->ext_mld_capa_and_ops)))
 					goto nla_put_failure;
 
 				nla_nest_end(msg, nested_ext_capab);
-- 
2.45.2

